{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lab 4: You are what you eat (RNA-seq)\n",
    "\n",
    "<font color=\"red\">This notebook is not graded, it is provided for informational purposes, but please do read through it. Note the paths are outdated (from a previous year where this was lab 3). The preprocessed files are now stored in `~/public/lab4/preprocessing/`.</font>\n",
    "\n",
    "This notebook contains background information about how the raw RNA-seq data was preprocessed to obtain gene-level count information from the raw reads. \n",
    "\n",
    "\n",
    "# 0. Accessing the raw data\n",
    "\n",
    "Raw data files used for preprocessing were obtained using the code below (in the terminal). Please check out the links below, but do not run this! It is meant to show you how we did this, which you may find helpful as a guide when you need to find data for your own project. \n",
    "\n",
    "```\n",
    "# Accessions for each file were found at\n",
    "# https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE87565\n",
    "# and navigating to the SRA link for each file\n",
    "\n",
    "fastq-dump -Z SRR4340937 > Chow_Rep1.fq\n",
    "fastq-dump -Z SRR4340938 > Chow_Rep2.fq\n",
    "fastq-dump -Z SRR4340939 > Chow_Rep3.fq\n",
    "\n",
    "fastq-dump -Z SRR4340943 > HFD_Rep1.fq\n",
    "fastq-dump -Z SRR4340944 > HFD_Rep2.fq\n",
    "fastq-dump -Z SRR4340945 > HFD_Rep3.fq\n",
    "\n",
    "# Get reference genome from ENSEMBL\n",
    "wget -O GRCm38.fa http://labshare.cshl.edu/shares/gingeraslab/www-data/dobin/STAR/STARgenomes/ENSEMBL/mus_musculus/ENSEMBL.mus_musculus.release-75/Mus_musculus.GRCm38.75.dna.primary_assembly.fa\n",
    "\n",
    "# Get gene annotations from ENSEMBL\n",
    "wget -O GRCm38.75.gtf http://labshare.cshl.edu/shares/gingeraslab/www-data/dobin/STAR/STARgenomes/ENSEMBL/mus_musculus/ENSEMBL.mus_musculus.release-75/Mus_musculus.GRCm38.75.gtf\n",
    "\n",
    "```\n",
    "\n",
    "# 1. Alignment of RNA-seq data\n",
    "\n",
    "As for most NGS analyses we'll be doing (besides assembly), the first step is to align our reads so we know where they came from. As we discussed in class, we cannot simply align reads to the reference genome using something like BWA-MEM, since this does not take into account \"splice junctions\" that will be prevalent in our RNA-seq data.\n",
    "\n",
    "We used an RNA-seq aligner called STAR for this. STAR takes in fastq files and a reference *transcriptome* and outputs BAM files with aligned reads. **Because STAR can be very memory intensive (20+GB RAM), we have already run this for you. We will still walk through the steps to run it here as a reference for using this in the future.** \n",
    "\n",
    "First, similar to BWA-MEM, we need to create an \"index\" that will be used during alignment to rapidly look up where each read came from. We created an index using the following command:\n",
    "\n",
    "```shell\n",
    "REFFA=/datasets/cs185-sp22-a00-public/genomes/GRCm38.fa\n",
    "GTF=/datasets/cs185-sp22-a00-public/genomes/GRCm38.75.gtf\n",
    "STARDIR=/datasets/cs185-sp22-a00-public/lab3/preprocessing/STAR\n",
    "STAR \\\n",
    "    --runMode genomeGenerate \\\n",
    "    --genomeDir ${STARDIR} \\\n",
    "    --genomeFastaFiles ${REFFA} \\\n",
    "    --sjdbGTFfile ${GTF} \\\n",
    "    --sjdbOverhang 49\n",
    "```\n",
    "\n",
    "<blockquote>UNIX TIP: Note, in the above command we are taking advantage of bash variables. We can define variables using the syntax: VAR=value (important, no spaces!) then reference variables using \\$VAR or \\${VAR}.</blockquote>\n",
    "\n",
    "Input parameters to this command were:\n",
    "* `--runMode genomeGenerate`: tells STAR we're creating an index\n",
    "* `--genomeDir ${STARDIR}`: tells STAR where to put the output files\n",
    "* `--genomeFastaFiles ${REFFA}` pointed STAR to the fasta file for our reference genome. Here we are using a mouse genome with the GRCm38 build.\n",
    "* `--sjdbGTFfile ${GTF}` provides the gene annotations (GTF format). This file is critical for telling STAR where all the exon-exon boundaries are and where genes start and end. You should take a look at this file to get an idea of what these gene annotations look like. See the [GTF format spec](https://uswest.ensembl.org/info/website/upload/gff.html) and lecture slides for more details. This file was created by ENSEMBL build 75 (http://feb2014.archive.ensembl.org/Mus_musculus/Info/Index).\n",
    "* `--sjdbOverhang` should be set to the read length -1 according to the STAR manual.\n",
    "\n",
    "After creating the index, we can align reads to the transcriptome. We used the following bash commands to run the alignment:\n",
    "\n",
    "\n",
    "```shell\n",
    "OUTDIR=/datasets/cs185-sp22-a00-public/lab3/preprocessing\n",
    "\n",
    "# STAR options recommended by ENCODE\n",
    "STAROPTS=\"--outSAMattributes NH HI AS NM MD \\\n",
    "\t--outFilterType BySJout \\\n",
    "\t--outFilterMultimapNmax 20 \\\n",
    "\t--outFilterMismatchNmax 999 \\\n",
    "\t--outFilterMismatchNoverReadLmax 0.04 \\\n",
    "\t--alignIntronMin 20 \\\n",
    "\t--alignIntronMax 1000000 \\\n",
    "\t--alignMatesGapMax 1000000 \\\n",
    "\t--alignSJoverhangMin 8 \\\n",
    "\t--alignSJDBoverhangMin 1 \\\n",
    "\t--sjdbScore 1 \\\n",
    "\t--limitBAMsortRAM 50000000000\"\n",
    "\n",
    "for f in Chow_Rep1 Chow_Rep2 Chow_Rep3 HFD_Rep1 HFD_Rep2 HFD_Rep3\n",
    "do\n",
    "    STAR \\\n",
    "\t--runThreadN 5 \\\n",
    "\t--genomeDir ${STARDIR} \\\n",
    "\t--readFilesIn ${OUTDIR}/fastqs/${f}.fq \\\n",
    "\t--outFileNamePrefix ${OUTDIR}/${f} \\\n",
    "\t--outSAMtype BAM SortedByCoordinate \\\n",
    "\t--quantMode TranscriptomeSAM ${STAROPTS}\n",
    "    # Reorganize the output files\n",
    "    mv ${OUTDIR}/${f}Aligned.toTranscriptome.out.bam ${OUTDIR}/txBams/\n",
    "    mv ${OUTDIR}/${f}Aligned.sortedByCoord.out.bam ${OUTDIR}/genomeBams/\n",
    "    samtools index ${OUTDIR}/genomeBams/${f}Aligned.sortedByCoord.out.bam\n",
    "done\n",
    "```\n",
    "\n",
    "<blockquote>UNIX tip: The for loop goes through each of our samples and runs a separate command for each, similar to for loops in Python. You may find the for loop syntax useful for running additional commands below. See more about bash for loops: https://www.cyberciti.biz/faq/bash-for-loop/.</blockquote>\n",
    "\n",
    "This command uses STAR to align each of our fastq files to the reference genome and reference transcriptome. \n",
    "\n",
    "This outputs two BAM files:\n",
    "* `${f}Aligned.sortedByCoord.out.bam`: contains our reads aligned to the reference mouse *genome* (GRChm38).\n",
    "* `${f}Aligned.toTranscriptome.out.bam`: contains our reads aligned to the mouse *transcriptome* (Ensembl version 75).\n",
    "\n",
    "If you do `samtools view` on these files, you will see the first BAM files (genome) are aligned to chromosomes like we are used to (e.g. 1, 2, etc.). The second BAM files (transcriptome) are aligned to *transcripts* (named things like \"ENSMUST00000074245\"). These transcriptome BAMs are required for the tools in the next section for quantifying gene expression.\n",
    "\n",
    "A major difference between the two alignments is that alignments of RNA-seq reads to the reference genome will contain large gaps due to splice junctions. Find reads with such gaps by looking for CIGAR scores containing an \"N\" character (gap). e.g.:\n",
    "```\n",
    "samtools view /datasets/cs185-sp22-a00-public/lab3/preprocessing/genomeBams/Chow_Rep1Aligned.sortedByCoord.out.bam | awk '($6 ~ /N/)' | head\n",
    "```\n",
    "You should see reads with CIGAR scores like `38M723N12M`. This means a read matched for 38bp on one exon, spanned a 723bp intron, and matched to 12bp on the next exon. You won't typically see this type of read in the transcriptome BAMs, since in those cases reads were aligned directly to transcripts with intron sequences removed.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Quantifying gene expression\n",
    "\n",
    "Next, we will use RSEM, a tool for quantifying gene expression from RNA-seq. It takes in reads aligned to the transcriptome, which we obtained using STAR, and outputs estimated expression levels of each gene.\n",
    "\n",
    "RSEM requires an initial step to preprocess the reference transcriptome, similar to the index step required for sequence alignment. We have run this for you using the command below:\n",
    "\n",
    "```shell\n",
    "REFFA=/datasets/cs185-sp22-a00-public/genomes/GRCm38.fa\n",
    "GTF=/datasets/cs185-sp22-a00-public/genomes/GRCm38.75.gtf\n",
    "RSEMOUT=/datasets/cs185-sp22-a00-public/lab3/preprocessing/RSEM/RSEM\n",
    "rsem-prepare-reference \\\n",
    "    --gtf ${GTF} ${REFFA} ${RSEMOUT}\n",
    "```\n",
    "\n",
    "This will generate multiple index files with prefix `/datasets/cs185-sp22-a00-public/lab3/RSEM/RSEM*`.\n",
    "\n",
    "Now, we are ready to run RSEM for expression quantification. The following command shows how to run RSEM on a single sample:\n",
    "\n",
    "```shell\n",
    "rsem-calculate-expression \\\n",
    "\t-p 5 \\\n",
    "\t--fragment-length-mean -1 \\\n",
    "\t--seed-length 25 \\\n",
    "\t--bam /datasets/cs185-sp22-a00-public/lab3/preprocessing/txBams/Chow_Rep1Aligned.toTranscriptome.out.bam \\\n",
    "\t${RSEMOUT} \\\n",
    "\t/datasets/cs185-sp22-a00-public/lab3/Chow_Rep1\n",
    "```\n",
    "(we also included `-p 5` to run on 5 processors to speed this up.)\n",
    "\n",
    "We ran `rsem-calculate-expression` on each of the six samples (separately) to create the following output files (for each condition. shown here for `Chow_Rep1`):\n",
    "* `Chow_Rep1.genes.results`: gene-level expression results\n",
    "* `Chow_Rep1.isoforms.results`: isoform (transcript)-level expression results\n",
    "\n",
    "We have included the `*.genes.results` for gene-level analysis, which you will use as a starting point in the lab assignment."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
