{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "547ed7f5095c3ce0dee20584bc7a66fa",
     "grade": false,
     "grade_id": "text-lab2-intro",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "source": [
    "# Lab 2: So where do all those reference genomes come from anyway? (Sequence assembly)\n",
    "\n",
    "This week's exercises consist of two parts:\n",
    "* **Part 1 (10 pts)**: We'll derive how to estimate the size of a genome based on kmer frequencies using a toy example. In the lab, we'll do this on a real genome.\n",
    "\n",
    "* **Part 2 (10 pts)**: Review questions to test your understanding of course material.\n",
    "\n",
    "Reminder: you may work with a partner or consult your classmates or the internet for help on the lab assignments!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "7cdd040a4e3eca5a2a8d33b3174ae973",
     "grade": false,
     "grade_id": "text-lab2-part1",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "\n",
    "# Part 1: Estimating genome size\n",
    "\n",
    "## 1.1 Intro to kmer counts and coverage\n",
    "\n",
    "First, let's consider kmers originating from $N$ copies of an example genome ($G$) of length $|G|=15$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "a24cb53c6199ccf44da2fa558643e769",
     "grade": false,
     "grade_id": "text-lab2-exgenome",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "outputs": [],
   "source": [
    "GENOME=\"ACTCGGCGTTACGCG\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "2541385e201842e8e79132c8a2d876e4",
     "grade": false,
     "grade_id": "kmer-num",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "source": [
    "**Question 1 (1 pt)**: How many unique kmers of length 7 can be generated from our genome? Set the variable `num_unique_kmers` to your answer. List the unique kmers in the variable `unique_kmer_list`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "b3e28cca68e8f7ea7d9879a086159ea4",
     "grade": false,
     "grade_id": "kmer-num-answer",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9\n",
      "['GTTACGC', 'CTCGGCG', 'TCGGCGT', 'TTACGCG', 'CGTTACG', 'GCGTTAC', 'CGGCGTT', 'ACTCGGC', 'GGCGTTA']\n"
     ]
    }
   ],
   "source": [
    "num_unique_kmers = 0\n",
    "unique_kmer_list = [\"NNNNNNN\",\"NNNNNNN\"] # Replace with your answer\n",
    "\n",
    "# your code here\n",
    "def generate_kmers(sequence, k):\n",
    "    kmers = []\n",
    "    for i in range(len(sequence) - k + 1):\n",
    "        kmer = sequence[i:i+k]\n",
    "        kmers.append(kmer)\n",
    "    return kmers\n",
    "\n",
    "k = 7\n",
    "kmers = generate_kmers(GENOME, k)\n",
    "unique_kmer_list = list(set(kmers))\n",
    "# num_unique_kmers = len(GENOME) - k +1\n",
    "num_unique_kmers = len(unique_kmer_list)\n",
    "\n",
    "print(num_unique_kmers)\n",
    "print(unique_kmer_list)\n",
    "# raise NotImplementedError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "20ef90cf2d2d4be00d6dba41d955043d",
     "grade": true,
     "grade_id": "kmer-num-test",
     "locked": true,
     "points": 1,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "assert(num_unique_kmers > 1 and num_unique_kmers < len(GENOME))\n",
    "assert(num_unique_kmers == len(unique_kmer_list))\n",
    "for val in unique_kmer_list:\n",
    "    assert(type(val)==str)\n",
    "    assert(len(val)==7)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "4560dabb9735616117c06f1edbff32a3",
     "grade": false,
     "grade_id": "text-kmer-genomesize",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "source": [
    "**Question 2 (2 pts)**: What would your answer to question 1 (number of unique 7-mers) be for a genome of length 100? \n",
    "1000? 1000000?\n",
    "\n",
    "Complete the function `GetNumUniqueKmers` below to return the number of unique kmers of length $k$ for a genome of size $|G|$. Ignore the possibility that a single kmer may be seen more than once in the genome. \n",
    "\n",
    "Notice that, as the genome size gets much bigger than the kmer size, the number of total kmers forms a good approximation to the genome size."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "baeea6e1c57ccb0507417f1a63e03017",
     "grade": false,
     "grade_id": "kmer-genomeisze-answer",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "def GetNumUniqueKmers(G, k):\n",
    "    \"\"\"Compute the number of unique kmers for a genome of size G\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    G : int\n",
    "       Size, in bp, of the target genome\n",
    "    k : int\n",
    "       Length of the kmer being considered\n",
    "       \n",
    "    Returns\n",
    "    -------\n",
    "    num_unique_kmers : int\n",
    "       Number of unique kmers possible\n",
    "    \"\"\"\n",
    "    num_unique_kmers = 0\n",
    "    # your code here\n",
    "    num_unique_kmers = G-k+1\n",
    "#     raise NotImplementedError\n",
    "    return num_unique_kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "7fe01d6bea25c703da0d9c68a8cfb1e5",
     "grade": true,
     "grade_id": "kmer-genomeisze-test",
     "locked": true,
     "points": 2,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "assert(GetNumUniqueKmers(100, 7)==94)\n",
    "assert(GetNumUniqueKmers(1000000, 7) == 999994)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "9d07fb627e7c7dbb9c831d43f0e99588",
     "grade": false,
     "grade_id": "text-kmer-cov",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "source": [
    "**Question 3 (1 pt)**: We decide to sequence our genome with 100 reads of length ($L$) 7bp. What will be the average coverage ($C$) of our length 15 genome above? Set the variable `meancov` below to your answer.\n",
    "\n",
    "Recall that *coverage* refers to the average number of times each base in the genome is sequenced. (see the formula for coverage from last week.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "cfb074fbd56f562352dcc085dfc2cba1",
     "grade": false,
     "grade_id": "kmer-cov-answer",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46.666666666666664\n"
     ]
    }
   ],
   "source": [
    "meancov = 0.0\n",
    "# your code here\n",
    "# readlength\n",
    "l=7\n",
    "# genomesize\n",
    "g=15\n",
    "# num reads\n",
    "n=100\n",
    "meancov = (l*n)/g\n",
    "print(meancov)\n",
    "# raise NotImplementedError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "5f3cfa9147490ed98d159c9e49add622",
     "grade": true,
     "grade_id": "kmer-cov-test",
     "locked": true,
     "points": 1,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "assert(meancov > 10.0 and meancov < 100.0)\n",
    "assert(type(meancov)==float)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "1becd8b117be85d7971f31b247252830",
     "grade": false,
     "grade_id": "text-kmercov",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "source": [
    "We can also consider the *kmer coverage* ($M$), or the average number of times each *kmer* of a given length is seen. For example, if we have reads of length $L=7$, and are using $k=7$, then we will sequence a single kmer for each read. If we know the total number of possible kmers (your answer to question 2), then we can compute $M$ as the total number of kmers seen divided by the total number of possible kmers. \n",
    "\n",
    "**Question 4 (1 pt)**: Continuing our example from question 3, with $N=100$ reads of length $L=7$, $|G|=15$, and $k=7$, what will be the average *kmer coverage*? Set the variable `kmer_cov1` to your answer below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "b7278f84d5fa63db1363ffa5008c4431",
     "grade": false,
     "grade_id": "kmercov-answer",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11.11111111111111\n"
     ]
    }
   ],
   "source": [
    "kmer_cov1 = 0\n",
    "# your code here\n",
    "# ((7-7+1)*100) = 100\n",
    "kmer_cov1 = 100/GetNumUniqueKmers (15, 7)\n",
    "print(kmer_cov1)\n",
    "# raise NotImplementedError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "921924195bc46133a71a9d27da3346e9",
     "grade": true,
     "grade_id": "kmerread-test-1",
     "locked": true,
     "points": 1,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "assert(kmer_cov1 > 10 and kmer_cov1 < 20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "88ada2b0b0fc41050bab1e588e3dad38",
     "grade": false,
     "grade_id": "kmeread",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "source": [
    "**Question 5 (1 pt)**: What about if read length $L=10$? How many kmers of length $k=7$ will we observe per read? What will be the average kmer coverage with 100 reads for a genome of length 15? Set the variable `kmer_cov2` to your answer below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "da451ccec4aece81c9472d2fa6a71622",
     "grade": false,
     "grade_id": "kmerread-answer",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44.44444444444444\n"
     ]
    }
   ],
   "source": [
    "kmer_cov2 = 0\n",
    "# your code here\n",
    "kmer_cov2 = ((10-7+1)*100)/GetNumUniqueKmers (15, 7)\n",
    "print(kmer_cov2)\n",
    "# raise NotImplementedError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "c0529e45453c7e3021fa3b479c114d80",
     "grade": true,
     "grade_id": "kmerread-test-2",
     "locked": true,
     "points": 1,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "assert(kmer_cov2 > 30 and kmer_cov2 < 50)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "9b529ad0ecae1602668b306955e9c068",
     "grade": false,
     "grade_id": "test-q6",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "source": [
    "Now let's write some code to check our answers.\n",
    "\n",
    "**Question 6 (4 pts)**: Complete the code below, which generates $N$ reads of length $L$ from an example genome $G$ and plots a histogram of kmer counts for a given kmer length $k$.\n",
    "\n",
    "You will need to fill in the functions:\n",
    "* `GetKmersFromReads`: which obtains a list of all kmers of a given length observed in the reads\n",
    "* `GetMeanKmerCov`: which computes the mean kmer coverage ($M$) based on the read length $L$, kmer size $k$, and number of reads $N$. \n",
    "\n",
    "The code will create a histogram of kmer counts. If your result for `GetMeanKmerCov` is correct it should be around the mean value of the histogram."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "45a32461663c558712c86700e228f227",
     "grade": false,
     "grade_id": "plot-khist",
     "locked": false,
     "schema_version": 3,
     "solution": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAboAAADQCAYAAABx76QNAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAAV/0lEQVR4nO3dfbBcdX3H8feHEBUE5CERaJLLBYxU0BrwNpCiEB6kECIg8qigMNYAgiUFpg0UQWVaoBWriEMmLRgYKUgN0ADJKDCEB5WQBwIhhECkqUQyIVAbSHjQwLd/nHPLsuzee+7ePffs2f28Zs6cs+fxk51NvjlPv58iAjMzs3a1WdEBzMzM8uRCZ2Zmbc2FzszM2poLnZmZtTUXOjMza2sudGZm1tY2LzrAQI0YMSK6u7uLjmE2NFasSMZ77FFsDrMWt2jRopciYmStZaUrdN3d3SxcuLDoGGZDY+LEZDxvXpEpzFqepP+ut8yXLs3MrK250JmZWVvL7dKlpA8ADwLvT4/zs4i4tGodAT8AJgGvAadFxOK8MpmVji9Zmg1anvfo3gQOjogNkoYDD0uaGxGPVKxzBDA2HfYFrk3HZmZmTZHbpctIbEg/Dk+H6hakjwZuTNd9BNhW0s55ZTIrne9+NxnMrGG5PnUpaRiwCPgI8KOImF+1yijg+YrPq9N5a6r2MwWYAtDV1ZVbXrNauqfd3fC2q644cnAHv+uuZHzBBYPbj1kHy/VhlIh4KyLGAaOB8ZI+XrWKam1WYz8zIqInInpGjqz5moSZmVlNQ/LUZUT8LzAPOLxq0WpgTMXn0cALQ5HJzMw6Q26FTtJISdum01sAhwJPV602G/iyEvsB6yNiDWZmZk2S5z26nYEb0vt0mwG3RsRdks4EiIjpwBySVwtWkrxecHqOeczKZ4stik5gVnq5FbqIeALYu8b86RXTAZydVwaz0ps7t+gEZqXnllHMzKytudCZtbLLLksGM2uYC51ZK7vvvmQws4a50JmZWVtzoTMzs7bmQmdmZm2tdD2Mm3WUHXYoOoFZ6bnQmbWyWbOKTmBWer50aWZmbc2FzqyVXXhhMphZw3zp0qyV/frXRScwKz2f0ZmZWVtzoTMzs7bmQmdmZm3N9+jMWtno0UUnMCs9FzqzVvaTnxSdwKz0fOnSzMzaWm6FTtIYSfdLWi5pmaRza6wzUdJ6SUvS4ZK88piV0tSpyWBmDcvz0uUm4PyIWCxpa2CRpHsi4qmq9R6KiMk55jArryVLik5gVnq5ndFFxJqIWJxOvwosB0bldTwzM7NahuQenaRuYG9gfo3FEyQ9LmmupL3qbD9F0kJJC9etW5dnVDMzazO5FzpJWwGzgKkR8UrV4sXALhHxSeCHwB219hERMyKiJyJ6Ro4cmWteMzNrL7kWOknDSYrcTRFxW/XyiHglIjak03OA4ZJG5JnJrFQ++tFkMLOG5fYwiiQB1wHLI+J7ddbZCVgbESFpPEnhfTmvTGalM2NG0QnMSi/Ppy73B04Flkpaks67COgCiIjpwHHAWZI2Aa8DJ0VE5JjJzMw6TG6FLiIeBtTPOtcA1+SVwaz0pkxJxj6zM2vYgAqdpO2AMRHxRE55zKzSM88UncCs9Pp9GEXSPEnbSNoeeBz4saSa99zMzMxaTZanLj+UvhZwLPDjiPgUcGi+sczMzJojS6HbXNLOwAnAXTnnMTMza6os9+i+DfwceDgiFkjaDXg231hmBsC4cUUnMCu9PgudpGEkD5/8We+8iHgO+ELewcwM+P73i05gVnp9XrqMiLeAo4Yoi5mZWdNluXT5K0nXAD8FNvbO7O2ZwMxydMopydg9jZs1LEuh+4t0/J2KeQEc3Pw4ZvYuq1cXncCs9PotdBFx0FAEMTMzy0OWF8Z3lHSdpLnp5z0lfTX/aGZmZoOX5T26mSSvF/xJ+vkZYGpOeczMzJoqS6EbERG3Am8DRMQm4K1cU5lZYsKEZDCzhmV5GGWjpB1IHkBB0n7A+lxTmVni8suLTmBWelkK3XnAbGB3Sb8ERpL0I2dmZtbysjx1uVjSgcAeJP3LrYiIP+aezMzgC2kjRLNmFZvDrMT6LXRpM2CTgO50/cMkERHuqscsby+/XHQCs9LL8jDKncBpwA7A1hVDnySNkXS/pOWSlkk6t8Y6knS1pJWSnpC0zwDzm5mZ9SnLPbrRlY06D8Am4Pz00ufWwCJJ90TEUxXrHAGMTYd9gWvTsZmZWVNkOaObK+mwge44Itb0tocZEa8Cy4FRVasdDdwYiUeAbdO+78zMzJoiyxndI8DtkjYD/kjyQEpExDZZDyKpG9gbmF+1aBTwfMXn1em8NVXbTwGmAHR1dWU9rFn5HXJI0QlKp3va3Q1vu+qKI5uYxFpFlkJ3FTABWBoRMdADSNoKmAVMjYhXqhfX2OQ9x4iIGcAMgJ6engFnMCutb36z6ARmpZfl0uWzwJMNFrnhJEXupoi4rcYqq4ExFZ9HAy8M9DhmZmb1ZDmjWwPMSxt1frN3Zn+vF0gScB2wvI91ZwPnSLqF5CGU9RGxps66Zp3niCOS8dy5xeYwK7Eshe6/0uF96QA1Li/WsD9wKrBU0pJ03kVAF0BETAfmkLyjtxJ4DTg9a3CzjvD660UnMCu9LIXuhohYVTlD0p/3t1FEPEzte3CV6wRwdoYMZmZmDclyj26WpP9/LUDSAcD1+UUyMzNrniyF7kzgDkk7SZoEXE1yudHMzKzlZWnUeYGkvwZ+AbwBfDYi1uWezMxg8uSiE5iVXt1CJ+lO3v3QyZYk/dBdlzbqfFTe4cw63gUXFJ3ArPT6OqP77pClMDMzy0ndQhcRDwxlEDOrYeLEZDxvXpEpzEoty8MoZmZmpeVCZ2Zmba1uoZN0Xzq+cujimJmZNVdfD6PsLOlA4Ki0Lcp3tXLS29ecmZlZK+ur0F0CTCPpUaC6UeYADs4rlJmlTjih6ARmpdfXU5c/A34m6ZsRcdkQZjKzXl//etEJzEovS8sol0k6CjggnTUvIu7KN5aZAfDaa8l4yy2LzWFWYv0WOkmXA+OBm9JZ50raPyIuzDWZmcGktFlZv0dn1rAs3fQcCYyLiLcBJN0APAa40JmZWcvL+h7dthXTH8ohh5mZWS6ynNFdDjwm6X6SVwwOwGdzZmZWEv2e0UXEzcB+wG3pMCEibulvO0nXS3pR0pN1lk+UtF7SknS4ZKDhzczM+pPljI6IWAPMHuC+ZwLXADf2sc5DEeEOt8zqOe20ohOYlV6mQteIiHhQUnde+zfrCC50ZoNWdKPOEyQ9LmmupL3qrSRpiqSFkhauW+fOza2DvPRSMphZw/osdJI2q3ePrQkWA7tExCeBHwJ31FsxImZERE9E9IwcOTKnOGYt6LjjksHMGtZnoUvfnXtcUlezDxwRr0TEhnR6DjBc0ohmH8fMzDpblnt0OwPLJD0KbOydGRFHDebAknYC1kZESBpPUnRfHsw+zczMqmUpdN9uZMeSbgYmAiMkrQYuBYYDRMR04DjgLEmbgNeBkyIiGjmWmZlZPVkadX5A0i7A2Ii4V9KWwLAM253cz/JrSF4/MDMzy02WRp2/BkwBtgd2B0YB04FD8o1mZpx1VtEJzEovy6XLs0l6L5gPEBHPSvpwrqnMLHHiiUUnMCu9LO/RvRkRf+j9IGlzkh7GzSxvzz+fDGbWsCxndA9IugjYQtJnga8Dd+Yby8wAOPXUZOz+6MwaluWMbhqwDlgKnAHMAS7OM5SZmVmzZHnq8u20s9X5JJcsV/g1ADMzK4ssT10eSfKU5W9I+qPbVdIZETE373BmZmaDleUe3VXAQRGxEkDS7sDdgAudmZm1vCyF7sXeIpd6DngxpzxmVun884tOYFZ6dQudpGPTyWWS5gC3ktyjOx5YMATZzOxznys6gVnp9XVGV/k3bC1wYDq9Dtgut0Rm9o4VK5LxHnsUm8OsxOoWuog4fSiDmFkNZ5yRjP0enVnDsjx1uSvwDaC7cv3BdtNjZmY2FLI8jHIHcB1Jayhv55rGzMysybIUujci4urck5iZmeUgS6H7gaRLgV8Ab/bOjIjFuaUyMzNrkiyF7hPAqcDBvHPpMtLPZpani92srNlgZSl0nwd2q+yqJwtJ1wOTSV44/3iN5QJ+AEwCXgNO81miWZVDDy06gVnpZem94HFg2wb2PRM4vI/lRwBj02EKcG0DxzBrb0uWJIOZNSzLGd2OwNOSFvDue3R9vl4QEQ9K6u5jlaOBG9OeEB6RtK2knSNiTYZMZp1h6tRk7PfozBqWpdBdmtOxRwGVXSevTue9p9BJmkJy1kdXV1dOcaxddU+7u7THvuW5lwE4qYA/w6orjhzyY5rlIUt/dA/kdGzVOlydDDOAGQA9PT3uC8/MzDLL0jLKq7xTgN4HDAc2RsQ2gzz2amBMxefRwAuD3KeZmdm7ZDmj27rys6RjgPFNOPZs4BxJtwD7Aut9f87MzJotyz26d4mIOyRN6289STcDE4ERklaT3Osbnu5jOjCH5NWClSSvF7gRabMq/3TAV4qOYFZ6WS5dHlvxcTOghzr30ipFxMn9LA/g7P72Y9bJFo/+WNERzEovyxldZb90m4BVJK8GmFnO9lm9HHDBMxuMLPfofEnRrCB/++ANAJz0xSsKTmJWXnULnaRL+tguIuKyHPKYmZk1VV9ndBtrzPsg8FVgB8CFzszMWl7dQhcRV/VOS9oaOJfkychbgKvqbWdmZtZK+rxHJ2l74DzgS8ANwD4R8fuhCGZmZtYMfd2j+2fgWJKmtz4RERuGLJWZAfCdQ6YUHcGs9Po6ozufpLeCi4G/T7qPA5I2KqMJTYCZWT+e2nG3oiOYlV5f9+iy9FVnZjnaf9USAH7ZPa7QHGZlNuAmwMxs6HzjV7cALnRmg+GzNjMza2sudGZm1tZc6MzMrK250JmZWVvzwyhmLeyivzyn6AhmpedCZ9bCntthdNERzEov10uXkg6XtELSylq9kkuaKGm9pCXp0FePCWYd55CV8zlk5fyiY5iVWm5ndJKGAT8CPgusBhZImh0RT1Wt+lBETM4rh1mZfe3R2wG47yP7FpzErLzyPKMbD6yMiOci4g8kvR64Z3IzMxtSeRa6UcDzFZ9Xp/OqTZD0uKS5kvbKMY+ZmXWgPB9GUY15UfV5MbBLRGyQNAm4Axj7nh1JU4ApAF1dXU2OaWZm7SzPM7rVwJiKz6OBFypXiIhXerv/iYg5wHBJI6p3FBEzIqInInpGjhyZY2QzM2s3eZ7RLQDGStoV+B1wEvDFyhUk7QSsjYiQNJ6k8L6cYyazUvmbyecXHcGs9HIrdBGxSdI5wM+BYcD1EbFM0pnp8unAccBZkjYBrwMnRUT15U2zjrVmG1/BMBusXF8YTy9HzqmaN71i+hrgmjwzmJXZ5OUPAnDXxw4oOIlZebllFLMWdspjyf8TXejMGudGnc3MrK250JmZWVtzoTMzs7bmQmdmZm3ND6OYtbCzjrmw6AhmpedCZ9bCfr/lh4qOYFZ6vnRp1sKOW3ovxy29t+gYZqXmQmfWwlzozAbPhc7MzNqaC52ZmbU1FzozM2trLnRmZtbW/HqBWQs77fhvFR3BrPRc6Mxa2BvDP1B0BLPS86VLsxZ2yuK7OWXx3UXHMCs1FzqzFjb56YeY/PRDRccwK7VcC52kwyWtkLRS0rQayyXp6nT5E5L2yTOPmZl1ntwKnaRhwI+AI4A9gZMl7Vm12hHA2HSYAlybVx4zM+tMeZ7RjQdWRsRzEfEH4Bbg6Kp1jgZujMQjwLaSds4xk5mZdZg8C90o4PmKz6vTeQNdx8zMrGF5vl6gGvOigXWQNIXk0ibABkkrBpkNYATwUhP200n8nQ3coL6zCb0TV05uSpiB0JVDfshKhfzWCv4zD1an//3cpd6CPAvdamBMxefRwAsNrENEzABmNDOcpIUR0dPMfbY7f2cD5++sMf7eBs7fWX15XrpcAIyVtKuk9wEnAbOr1pkNfDl9+nI/YH1ErMkxk5mZdZjczugiYpOkc4CfA8OA6yNimaQz0+XTgTnAJGAl8Bpwel55zMysM+XaBFhEzCEpZpXzpldMB3B2nhn60NRLoR3C39nA+TtrjL+3gfN3VoeSWmNmZtae3ASYmZm1tY4rdP01S2a1SVolaamkJZIWFp2nFUm6XtKLkp6smLe9pHskPZuOtysyY6up8519S9Lv0t/aEkmTiszYaiSNkXS/pOWSlkk6N53v31odHVXoMjZLZvUdFBHj/AhzXTOBw6vmTQPui4ixwH3pZ3vHTN77nQH8S/pbG5fe67d3bALOj4iPAfsBZ6f/jvm3VkdHFTqyNUtm1pCIeBD4n6rZRwM3pNM3AMcMZaZWV+c7sz5ExJqIWJxOvwosJ2lRyr+1Ojqt0LnJscYF8AtJi9KWaiybHXvfDU3HHy44T1mck/Zocr0vwdUnqRvYG5iPf2t1dVqhy9TkmNW0f0TsQ3LZ92xJBxQdyNrWtcDuwDhgDXBVoWlalKStgFnA1Ih4peg8razTCl2mJsfsvSLihXT8InA7yWVg69/a3h450vGLBedpeRGxNiLeioi3gX/Fv7X3kDScpMjdFBG3pbP9W6uj0wpdlmbJrIqkD0rauncaOAx4su+tLDUb+Eo6/RXgPwvMUgpVXXV9Hv/W3kWSgOuA5RHxvYpF/q3V0XEvjKePKn+fd5ol+4diE7U+SbuRnMVB0prOv/t7ey9JNwMTSVqRXwtcCtwB3Ap0Ab8Fjo8IP3yRqvOdTSS5bBnAKuAMt4H7DkmfBh4ClgJvp7MvIrlP599aDR1X6MzMrLN02qVLMzPrMC50ZmbW1lzozMysrbnQmZlZW3OhMzOztuZCZ9ZEkjZUTE9KW5LvKjLTYEka5x4ErMxc6MxyIOkQ4IfA4RHx2ybve1gz95fBOMCFzkrLhc6sySR9hqTpqiMj4jfpvJmSrk37EXtO0oFpg8XLJc2s2PYwSb+WtFjSf6TtGfb2B3iJpIeB46uOt6Ok2yU9ng5/kc4/T9KT6TA1nddd1ffbBZK+lU7Pk3SlpEclPSPpM2kLQt8BTkz7hjsxv2/OLB+bFx3ArM28n6TppYkR8XTVsu2Ag4GjgDuB/YG/AhZIGkfSFuvFwKERsVHS3wHnkRQagDci4tM1jnk18EBEfD4929tK0qeA04F9SRozny/pAeD3/eTfPCLGp5cqL42IQyVdAvRExDkD+B7MWobP6Mya64/Ar4Cv1lh2ZyRNES0F1kbE0rTh4mVAN0knmnsCv5S0hKS9wl0qtv9pnWMeTNLiP2ljyOuBTwO3R8TGiNgA3AZ8JkP+3gaCF6WZzErPZ3RmzfU2cAJwr6SLIuIfK5a9WbHOm1XbbA68BdwTESfX2ffGAeSo1SUVJL1TV/4H9wNVy3tzvYX/fbA24TM6syaLiNeAycCXJNU6s6vnEWB/SR8BkLSlpI9m2O4+4Kx0m2GStgEeBI5J9/FBkl4AHiJpOPnDknaQ9P40Z39eBbYewJ/DrKW40JnlIG01/nDgYklHZ9xmHXAacLOkJ0gK359m2PRc4CBJS0kuOe4VEYuBmcCjJK3a/1tEPBYRfyS55zcfuAuovo9Yy/3Ann4YxcrKvReYmVlb8xmdmZm1NRc6MzNray50ZmbW1lzozMysrbnQmZlZW3OhMzOztuZCZ2Zmbc2FzszM2tr/AaVOTdMN5M43AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 504x216 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "%pylab inline\n",
    "import numpy as np\n",
    "\n",
    "GENOME=\"ACTCGGCGTTACGCG\"\n",
    "k = 7\n",
    "N = 100\n",
    "L = 7\n",
    "\n",
    "def GetReads(GENOME, N, L):\n",
    "    \"\"\" Return a list of N reads of length L from a GENOME string\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    GENOME : str\n",
    "       String consisting of the genome sequence\n",
    "    N : int\n",
    "       number of reads\n",
    "    L : int\n",
    "       length (in bp) of the reads\n",
    "       \n",
    "    Returns\n",
    "    -------\n",
    "    reads : list of str\n",
    "       List of reads (consisting of strings of nucleotides)\n",
    "    \"\"\"\n",
    "    reads = []\n",
    "    for i in range(N):\n",
    "        start = np.random.randint(0, len(GENOME)-L+1)\n",
    "        reads.append(GENOME[start:start+L])\n",
    "    return reads\n",
    "\n",
    "def GetKmersFromReads(reads, k):\n",
    "    \"\"\" Return a list of all kmers of length k from a list of reads\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    reads : list of str\n",
    "       List of reads (consisting of strings of nucleotides)\n",
    "    k : int\n",
    "       K-mer length\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    kmers : list of str\n",
    "       List of kmers (consisting of strings of nucleotides length k)\n",
    "    \"\"\"\n",
    "    kmers = []\n",
    "    for read in reads:\n",
    "        for i in range(len(read)-k+1):\n",
    "            # Get the kmer starting at index i and add to the list of kmers\n",
    "            # your code here\n",
    "            kmer = read[i:i+k]\n",
    "            kmers.append(kmer)            \n",
    "#           raise NotImplementedError\n",
    "    return kmers\n",
    "\n",
    "def GetKmerCounts(kmer_list):\n",
    "    \"\"\" Count the number of times each unique kmer occurs.\n",
    "    \n",
    "    Should return a dictionary with kmer_counts[kmer] = number of times the \n",
    "    kmer is seen in kmer_list\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    kmer_list : list of str\n",
    "       List of kmers (consisting of strings of nucleotides length k)\n",
    "       \n",
    "    Returns\n",
    "    -------\n",
    "    kmer_counts : dict of str->int\n",
    "       Dictionary indicating the number of times each kmer is seen in kmer_list.\n",
    "       Key: kmer. Value: number of times seen\n",
    "    \"\"\"\n",
    "    kmer_counts = {}\n",
    "    for kmer in kmer_list:\n",
    "        kmer_counts[kmer] = kmer_counts.get(kmer, 0) + 1\n",
    "    return kmer_counts\n",
    "\n",
    "def GetMeanKmerCov(GENOME, N, k, L):\n",
    "    \"\"\"Compute the mean kmer coverage of a genome\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    GENOME : str\n",
    "       String consisting of the genome sequence\n",
    "    N : int\n",
    "       number of reads\n",
    "    k : int\n",
    "       K-mer length\n",
    "    L : int\n",
    "       length (in bp) of the reads\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    mean_kmer_cov : float\n",
    "       Mean number of times each kmer is seen\n",
    "    \"\"\"\n",
    "    num_kmers_seq = 0 # Set this below\n",
    "    num_kmers_uniq = -1 # Set this below\n",
    "    # Set num_kmers_seq and num_kmers_uniq based on the formulas above\n",
    "    # num_kmers_seq is the total number of kmers sequenced\n",
    "    # num_kmers_uniq is the number of unique kmers in GENOME (assuming all are unique)\n",
    "    # your code here\n",
    "    num_kmers_seq = N * (L - k + 1)\n",
    "    num_kmers_uniq = len(GENOME) - k + 1\n",
    "\n",
    "    #raise NotImplementedError\n",
    "    mean_kmer_cov = num_kmers_seq*1.0/num_kmers_uniq\n",
    "    return mean_kmer_cov\n",
    "\n",
    "def PlotKmerHist(kmer_counts, meanval=None):\n",
    "    \"\"\" Plot a histogram of kmer counts\n",
    "    \n",
    "    Returns a plot where the x-axis is kmer count, and the y-axis\n",
    "    is the number of kmers seen that many times.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    kmer_counts : dict of str->int\n",
    "       Dictionary indicating the number of times each kmer is seen in kmer_list.\n",
    "       Key: kmer. Value: number of times seen\n",
    "    meanval : float (optional)\n",
    "       Mean kmer coverage value\n",
    "    \"\"\"\n",
    "    fig = plt.figure()\n",
    "    fig.set_size_inches((7, 3))\n",
    "    ax = fig.add_subplot(111)\n",
    "    vals = list(kmer_counts.values())\n",
    "    bins = np.arange(0, max(vals)*1.5, 1)\n",
    "    ax.hist(vals, bins=bins)\n",
    "    if meanval is not None: ax.axvline(x=meanval, linestyle=\"dashed\", color=\"red\")\n",
    "    ax.set_xlabel(\"Kmer count\")\n",
    "    ax.set_ylabel(\"Number of kmers\")\n",
    "    \n",
    "reads = GetReads(GENOME, N, L)\n",
    "kmer_list = GetKmersFromReads(reads, k)\n",
    "kmer_counts = GetKmerCounts(kmer_list)\n",
    "PlotKmerHist(kmer_counts, meanval=GetMeanKmerCov(GENOME, N, k, L))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "20a3493fefd1172487806d11ccc876a5",
     "grade": true,
     "grade_id": "correct-plot-khist",
     "locked": true,
     "points": 2,
     "schema_version": 3,
     "solution": false
    }
   },
   "outputs": [],
   "source": [
    "\"\"\"Test GetKmersFromReads\"\"\"\n",
    "GENOME=\"ACTCGGCGTTACGCG\"\n",
    "k = 7\n",
    "N = 100\n",
    "L = 7\n",
    "reads = GetReads(GENOME, N, L)\n",
    "kmer_list = GetKmersFromReads(reads, k)\n",
    "assert(len(kmer_list)==N)\n",
    "\n",
    "k = 7\n",
    "N = 100\n",
    "L = 10\n",
    "reads = GetReads(GENOME, N, L)\n",
    "kmer_list = GetKmersFromReads(reads, k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "365aed3f684a7beb0cc88b197a25d651",
     "grade": true,
     "grade_id": "correct-plot-khist2",
     "locked": true,
     "points": 2,
     "schema_version": 3,
     "solution": false
    }
   },
   "outputs": [],
   "source": [
    "\"\"\"Test GetMeanKmerCov\"\"\"\n",
    "GENOME=\"ACTCGGCGTTACGCG\"\n",
    "k = 7\n",
    "N = 100\n",
    "L = 7\n",
    "\n",
    "k = 5\n",
    "N = 1000\n",
    "L = 10"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "f26416497f700567b8680827513aec0c",
     "grade": false,
     "grade_id": "text-larger-genomes",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "source": [
    "Now let's take a look at kmer histograms for larger genomes to get a feel for what this will look like in real data. Play around with the parameters in the cell below and see how they affect the histograms. You can also use this to make sure your estimate fo the mean kmer coverage is correct."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAbYAAADQCAYAAABr00SDAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAAVhUlEQVR4nO3dfbRVdZ3H8fdH0HwkUZEh0K6WozI5qd3UJNHEGkcIH0bMyhY6Fo2rCJ+WIuP41BrD0kxtljOkCWvZSEg+opl6R7B0QhE0VCSNiLAbXMt8oPGR7/yx960j3od97j377nP2+bzWuuucvc/Z+3z48fDl99u/89uKCMzMzMpis6IDmJmZ1ZILm5mZlYoLm5mZlYoLm5mZlYoLm5mZlYoLm5mZlcrgogNksdNOO0VLS0vRMSxvK1cmj3vuWWwOM2sIjz322AsRMWzT/Q1R2FpaWliyZEnRMSxvhx2WPC5cWGQKM2sQkn7T1X4PRZqZWam4sJmZWak0xFCkNQkPQZpZDbjHZmZmpeLCZvXj8suTHzOzfvBQpNWNn189B4ATX9gbgNUzxxcZx8walHtsZmZWKi5sZmZWKi5sZmZWKr7GZnXjtcHvKTqCmZWAC5vVjZNPuLjoCGZWAh6KNDOzUnFhs7ox9aGbmPrQTUXHMLMG58JmdWPMb55gzG+eKDqGmTU4FzYzMysVFzYzMysVFzYzMysVT/e3uvHiVkOKjmBmJeDCZnXjtGNnFB3BzEog16FISWdIekrSk5JukrSlpB0k3Sfp2fRxaJ4ZzMysueRW2CSNBL4GtEbEh4BBwInAdKAtIvYA2tJtM85ZNJtzFs0uOoaZNbi8hyIHA1tJehPYGvgdcB5wWPr6HGAhcG7OOawB7P/8M0VHMLMSyK3HFhHPA5cDa4B24KWIuBcYHhHt6XvagZ3zymBmZs0nz6HIocDRwG7A+4BtJJ1UxfFTJC2RtKSjoyOvmGZmVjJ5DkUeAfw6IjoAJN0CHAyskzQiItoljQDWd3VwRMwCZgG0trZGjjmtTrVMv+sd26tnji8oiZk1kjxnRa4BDpK0tSQB44AVwB3A5PQ9k4Hbc8xgDaR9u51o326nomOYWYPLrccWEYslzQeWAm8By0h6YNsC8ySdSlL8JuWVwRrLGZ8+u+gIZlYCuc6KjIgLgQs32f06Se/NzMys5rxWpNWNC+6fxQX3zyo6hpk1OC+pZXVj9PpVRUcwsxJwYbOGseksyUqeMWlmnTwUaWZmpeLCZmZmpeKhSKsbq3YYWXQEMysBFzarGzOOnFp0BDMrAQ9FmplZqbiwWd249J5ruPSea4qOYWYNrqqhyHTF/l0i4hc55bEmsun0/bl/fL6gJGZWJr322CQtlDRE0g7AE8ANkr6dfzQzM7PqZRmKfG9EvAwcB9wQER8huSWNmZlZ3clS2Aan9007AViQcx4zM7N+yXKN7WLgJ8DPIuJRSbsDz+Yby5rR0zvvXnQEMyuBHgubpEEkk0X+vnNfRKwC/invYNZ8LjliStERzKwEeixsEfG2pInAlQOUx0qsp0WMzcxqJctQ5MOSvgv8ENjQuTMiluaWyprSlXdeDvhO2mbWP1kK28Hp4yUV+wI4vPZxrJmNeOWFoiOYWQn0Wtgi4hMDEcTMzKwWsnxBe7ik6yX9ON0eLenU/KOZmZlVL8v32GaTTPd/X7r9S+D0nPKYmZn1S5bCtlNEzAM2AkTEW8DbuaayprR05F4sHblX0THMrMFlmTyyQdKOJBNGkHQQ8FKuqawpffPQk/t8bG9fJVg9c3yfz21mjSVLYTsTuAP4gKSHgGHA8bmmMjMz66MssyKXSjoU2BMQsDIi3sw9mTWda2+9FIDTjp1RcBIza2S9FrZ0Wa2jgJb0/Z+SRET41jVWU0P/7+WiI5hZCWQZirwTeA1YTjqBJCtJ2wPXAR8iuUb3z8BKklVMWoDVwAkR8WI15zUzM+tOlsI2qnIR5CpdBdwTEcdL2gLYGpgBtEXETEnTgenAuX08v5mZ2Ttkme7/Y0mfqvbEkoYAY4HrASLijYj4E3A0MCd92xzgmGrPbWZm1p0sPbafA7dK2gx4k2QCSUTEkF6O2x3oAG6Q9GHgMWAaMDwi2klO0i5p564OljQFmAKw6667Zvm1WIN76P0fLjqCmZVAlsJ2BfAxYHlERJXn3h+YGhGLJV1FMuyYSUTMAmYBtLa2VvO51qCuGfPZoiOYWQlkGYp8FniyyqIGsBZYGxGL0+35JIVunaQRAOnj+irPa2Zm1q0sPbZ2YGG6CPLrnTt7m+4fEb+X9FtJe0bESmAc8HT6MxmYmT7e3tfwVi6z510IwMknXFxwEjNrZFkK26/Tny3SH0iX18pgKvCDdEbkKuAUkl7ivPQOAWuASVUlttLa8q3Xe3+TmVkvshS2ORGxunKHpI9mOXlEPA60dvHSuCzHm5mZVSvLNbYfSRrZuSFpLPD9/CKZmZn1XZbC9i/AbZL+RtJRwNUkS2yZmZnVnSyLID8q6WvAvSRLa30yIjpyT2ZNp+0DBxQdwcxKoNvCJulO3jlJZGuS+7Bdny6CPDHvcNZcvnfgcUVHMLMS6KnHdvmApTAzM6uRbgtbRCwayCBmc/87WZjmxM/NLDiJmTWyLJNHzMzMGoYLm5mZlUq3hU1SW/p42cDFMTMz65+eJo+MkHQoMFHSXJLb1fxFRCzNNZmZmVkf9FTYLiC5zcwoYNMFjwM4PK9Q1pwW7HVI0RHMrAR6mhU5H5gv6d8i4usDmMma1I37jy86gpmVQJaVR74uaSIwNt21MCIW5BvLmtGWb74GwGubb1lwEjNrZL3OipT0DWAaf72X2rR0n1lNzb75ImbffFHRMcyswWW5bc14YN+I2AggaQ6wDDgvz2BmZmZ9kfV7bNtXPH9vDjnMzMxqIkuP7RvAMkkPkEz5H4t7a2ZmVqeyTB65SdJC4KMkhe3ciPh93sHMzMz6IkuPjYhoB+7IOYs1ufn7HFF0BDMrgUyFzWwguLCZWS14EWSrG0P//BJD//xS0THMrMH1WNgkbSbpyYEKY83t2tu+wbW3+SuSZtY/PRa29LtrT0jadYDymJmZ9UuWa2wjgKckPQJs6NwZERNzS2VmZtZHWQrbxbmnMDMzq5FeJ49ExCJgNbB5+vxRIPO92CQNkrRM0oJ0ewdJ90l6Nn0c2sfsZmZm75JlEeQvAfOB/0p3jQRuq+IzpgErKranA20RsQfQlm6bceN+R3HjfkcVHcPMGlyW6f5fAcYALwNExLPAzllOLmkUySLK11XsPhqYkz6fAxyTMauV3IK9x7Jg77G9v9HMrAdZCtvrEfFG54akwSR30M7iO8A5wMaKfcPTlUw6VzTJVCSt/Ea83MGIlzuKjmFmDS5LYVskaQawlaRPAjcDd/Z2kKQJwPqIeKwvwSRNkbRE0pKODv9j1wyuXHAFVy64ougYZtbgshS26UAHsBz4MnA3cH6G48YAEyWtBuYCh0u6EVgnaQRA+ri+q4MjYlZEtEZE67BhwzJ8nJmZWbbV/TemNxddTDIEuTIieh2KjIjzSG9vI+kw4OyIOEnSt4DJwMz08fY+pzczM9tEr4VN0njgP4Ffkdy2ZjdJX46IH/fxM2cC8ySdCqwBJvXxPGZmZu+S5QvaVwCfiIjnACR9ALgLyFzYImIhsDB9/gdgXLVBzczMsshS2NZ3FrXUKrq5LmbWH9874NiiI5hZCXRb2CQdlz59StLdwDySa2yTSFYfMauptg8eWHQEMyuBnnpsn654vg44NH3eAXgZLOtVy/S7qnr/7n9YC8CqHUflnmX1zPE1/wwzqw/dFraIOGUgg5hd+pPvAnDi52YWnMTMGlmWWZG7AVOBlsr3+7Y1ZmZWj7JMHrkNuJ5ktZGNPb/Vmlm1Q49mZnnIUthei4irc09iZmZWA1kK21WSLgTuBV7v3BkRme/JZmZmNlCyFLZ9gC8Ah/PXochIt81q5pqDTyw6gpmVQJbCdiywe+Wta8zy8FDLvkVHMLMSyLK6/xPA9jnnMGP0ulWMXreq6Bhm1uCy9NiGA89IepR3XmPzdH+rqQvaZgH+HpuZ9U+WwnZh7inMzMxqJMv92BYNRBAzM7NayLLyyCsksyABtgA2BzZExJA8g5mZmfVFlh7bdpXbko4BDsgrkNU3LyZsZvUuyzW2d4iI2yRNzyOMNbdvjp1cdAQzK4EsQ5HHVWxuBrTy16FJs5pZOmrvoiOYWQlk6bFV3pftLWA1cHQuaayp7b92BeACZ2b9k+Uam+/L1kSKvIZ2zoNzgIH5HpuvFZqVV7eFTdIFPRwXEfH1HPKYmZn1S089tg1d7NsGOBXYEXBhawK+x5qZNZpuC1tEXNH5XNJ2wDTgFGAucEV3x5mZmRWpx2tsknYAzgQ+D8wB9o+IFwcimJmZWV/0dI3tW8BxwCxgn4h4dcBSWcOo5VDlJeOm1OxcZta8euqxnUWymv/5wL9K6twvkskjXlLLaurp4bsXHcHMSqDb+7FFxGYRsVVEbBcRQyp+tstS1CTtIukBSSskPSVpWrp/B0n3SXo2fRxay1+QNa4xqx9nzOrHi45hZg0uy41G++ot4KyI2Bs4CPiKpNHAdKAtIvYA2tJtM6Y+PJepD88tOoaZNbiq14rMKiLagfb0+SuSVgAjSVYtOSx92xxgIXBuXjnMsvAXts3KI88e219IagH2AxYDw9Oi11n8du7mmCmSlkha0tHRMRAxzcysBHIvbJK2BX4EnB4RL2c9LiJmRURrRLQOGzYsv4BmZlYquRY2SZuTFLUfRMQt6e51kkakr48A1ueZwczMmktu19iUfD/gemBFRHy74qU7gMnAzPTx9rwyWGOZ8Q9fLTqCmZVAboUNGAN8AVgu6fF03wySgjZP0qnAGmBSjhmsgazacVTREcysBPKcFfkzki9zd2VcXp9rjWvcc4sBaPvggQUnMbNGlmePzawqX3rkVsCFzcz6Z0Cm+5uZmQ0UFzYzMysVFzYzMysVFzYzMysVTx6xunHGhLOKjmBmJeDC1uRqeaPQ/mofUj9Lp3lRZLPG5aFIqxsTVjzIhBUPFh3DzBqce2xWN05adjcAC/YeW3ASM2tk7rGZmVmpuLCZmVmpuLCZmVmp+BqbWZV6mzHpGZVmxXJhy4n/caveacecV3QEMysBFzarGy9u/d6iI5hZCbiwNZl6+kL2po5ffj8A8/c5ouAk79ZTu9Vzm5o1I08esbpx/PL7/1LczMz6yj22BuTrd+Xi30+z2nJhKzkPkxXPhctsYHko0szMSsU9tm74f9lmZo3Jhc3qxsmTLio6gpmVgAtbA2iW62Svbb5l0REGRG+/n5Wve6TArHq+xmZ146Sld3HS0uYo4maWH/fY6kCte2SN2sOb8MxPAbhxf/dSzKzvCilsko4ErgIGAddFxMwicvRHtcXDw0tmZgNjwIciJQ0C/gP4R2A08FlJowc6h5mZlVMRPbYDgOciYhWApLnA0cDTAx2kml5XLYf3PPRo1rPe/kx71MN6UsTkkZHAbyu216b7zMzM+q2IHpu62BfvepM0BZiSbr4qaWUNPnsn4IUanKeZDFibfazzyWUTBuLj8lSzNtNltThLQ6iqzZqoXXrif8/g/V3tLKKwrQV2qdgeBfxu0zdFxCxgVi0/WNKSiGit5TnLzm1WPbdZ9dxm1XObda+IochHgT0k7SZpC+BE4I4CcpiZWQkNeI8tIt6S9FXgJyTT/b8fEU8NdA4zMyunQr7HFhF3A3cX8NE1HdpsEm6z6rnNquc2q57brBuKeNe8DTMzs4bltSLNzKxUmqKwSTpS0kpJz0maXnSeeiRpF0kPSFoh6SlJ09L9O0i6T9Kz6ePQorPWG0mDJC2TtCDddpv1QtL2kuZLeib9M/cxt1vPJJ2R/t18UtJNkrZ0m3Wt9IXNS3hl9hZwVkTsDRwEfCVtp+lAW0TsAbSl2/ZO04AVFdtus95dBdwTEXsBHyZpP7dbNySNBL4GtEbEh0gm3p2I26xLpS9sVCzhFRFvAJ1LeFmFiGiPiKXp81dI/qEZSdJWc9K3zQGOKSRgnZI0ChgPXFex223WA0lDgLHA9QAR8UZE/Am3W28GA1tJGgxsTfL9X7dZF5qhsHkJrypJagH2AxYDwyOiHZLiB+xcYLR69B3gHGBjxT63Wc92BzqAG9Ih3OskbYPbrVsR8TxwObAGaAdeioh7cZt1qRkKW6YlvCwhaVvgR8DpEfFy0XnqmaQJwPqIeKzoLA1mMLA/cG1E7AdswENoPUqvnR0N7Aa8D9hG0knFpqpfzVDYMi3hZSBpc5Ki9oOIuCXdvU7SiPT1EcD6ovLVoTHAREmrSYa4D5d0I26z3qwF1kbE4nR7Pkmhc7t17wjg1xHRERFvArcAB+M261IzFDYv4ZWBJJFc81gREd+ueOkOYHL6fDJw+0Bnq1cRcV5EjIqIFpI/V/8TESfhNutRRPwe+K2kPdNd40huW+V2694a4CBJW6d/V8eRXAd3m3WhKb6gLekokmshnUt4/XuxieqPpI8DPwWW89frRTNIrrPNA3Yl+cs1KSL+WEjIOibpMODsiJggaUfcZj2StC/JhJstgFXAKST/0Xa7dUPSxcBnSGYwLwO+CGyL2+xdmqKwmZlZ82iGoUgzM2siLmxmZlYqLmxmZlYqLmxmZlYqLmxmZlYqLmxm/SDp1YrnR6WrrO9aZKb+krRv+hUZs4bkwmZWA5LGAdcAR0bEmhqfe1Atz5fBvoALmzUsFzazfpJ0CPA9YHxE/CrdN1vStek97lZJOlTS99N7j82uOPZTkv5X0lJJN6drdSJptaQLJP0MmLTJ5w2XdKukJ9Kfg9P9Z6b36npS0unpvhZJT1Yce7aki9LnCyVdJukRSb+UdEi6Os8lwGckPS7pM/m1nFk+BhcdwKzBvYdkGaPDIuKZTV4bChwOTATuJFlb8ovAo+nKG2uB84EjImKDpHOBM0kKC8BrEfHxLj7zamBRRByb9ua2lfQRktU7DiRZ+HuxpEXAi73kHxwRB6RDjxdGxBGSLiC579dXq2gHs7rhHptZ/7wJPAyc2sVrd0aytM9yYF1ELI+IjcBTQAvJDV1HAw9Jepxkrb/3Vxz/w24+83DgWoCIeDsiXgI+DtwaERsi4lWSRXIPyZC/c7Hrx9JMZg3PPTaz/tkInADcL2lGRFxa8drrFe95fZNjBgNvA/dFxGe7OfeGKnJ0dXsmSNYVrPwP7JabvN6Z623874GVhHtsZv0UEX8GJgCfl9RVz607PwfGSPogQLpy+99mOK4NOC09ZlB6R+oHgWPSc2wDHEuyqPU6YGdJO0p6T5qzN68A21Xx6zCrKy5sZjWQrqh+JHC+pKMzHtMBnAzcJOkXJIVurwyHTgM+IWk5yRDi30XEUmA28AjJHRmui4hl6b27Lkn3LQA2vQ7YlQeA0Z48Yo3Kq/ubmVmpuMdmZmal4sJmZmal4sJmZmal4sJmZmal4sJmZmal4sJmZmal4sJmZmal4sJmZmal8v+AsAQZWZDNKgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 504x216 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Generate a random genome of 1kb\n",
    "nucs = [\"A\",\"C\",\"G\",\"T\"]\n",
    "genome_size = 1000\n",
    "GENOME = ''.join([nucs[np.random.randint(0, 4)] for  i in range(genome_size)])\n",
    "\n",
    "# Set real-world kmer size and read length\n",
    "k = 10\n",
    "L = 36\n",
    "N = 1000\n",
    "\n",
    "# GENOME=\"ACTCGGCGTTACGCG\"\n",
    "# k = 7\n",
    "# N = 100\n",
    "# L = 7\n",
    "\n",
    "\n",
    "reads = GetReads(GENOME, N, L)\n",
    "kmer_list = GetKmersFromReads(reads, k)\n",
    "kmer_counts = GetKmerCounts(kmer_list)\n",
    "PlotKmerHist(kmer_counts, meanval=GetMeanKmerCov(GENOME, N, k, L))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "ce3429dc2063348596a2cdeddea996bc",
     "grade": false,
     "grade_id": "text-part2-genomesize",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "source": [
    "## 1.2. Estimating genome size from kmer histograms\n",
    "\n",
    "We'd like to be able to estimate the size of a genome just based on the observed reads. It turns out we can leverage kmer histograms to solve this problem.\n",
    "\n",
    "After completing the exercise above, we now know how to compute *kmer coverage* ($M$: average number of times each kmer is seen) and *sequencing coverage* ($C$: average number of times each base pair is sequenced) as a function of:\n",
    "\n",
    "* $G$: the size of the genome\n",
    "* $N$: the number of reads\n",
    "* $L$: the length of the read\n",
    "* $k$: the size of the kmer\n",
    "\n",
    "Rearrange your formula for kmer coverage ($M$) to solve for $G$. Note, since in most settings $G-k+1$ is very close to $G$, you can replace wherever that term occurs in your formula with $G$ to simplify. That is, assume $G-k+1 \\approx G$.\n",
    "\n",
    "**Question 7 (1 pt)**: Complete the function `GetGenomeSize` below which computes genome size $G$ as a function of number of reads ($N$), read length ($L$), kmer size ($k$), and observed peak location in the kmer histogram ($M$) based on the formula you derive.\n",
    "\n",
    "Based on the observed kmer histogram above (for a genome of length 1000 with kmer size 10 and 1,000 reads of length 36), you should hopefully see a peak for $M$ around 25-30. If your formula is correct, you should estimate the genome size to be 900-1,080 which is pretty close to the true size of 1,000."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "385c07d343129f7c1e1f9758f3253c66",
     "grade": false,
     "grade_id": "genomesize-answer",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "def GetGenomeSize(N, k, L, M):\n",
    "    \"\"\" Estimate the size of a genome based on observed reads\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    N : int\n",
    "       number of reads\n",
    "    k : int\n",
    "       K-mer length\n",
    "    L : int\n",
    "       length (in bp) of the reads\n",
    "    M : float\n",
    "       Estimated mean kmer coverage.\n",
    "       \n",
    "    Return\n",
    "    ------\n",
    "    est_genome_size : int\n",
    "       Estimated genome size (in bp)\n",
    "    \"\"\"\n",
    "    est_genome_size = 0\n",
    "    # your code here\n",
    "    est_genome_size = (N*(L-k+1))/M\n",
    "    \n",
    "#     raise NotImplementedError\n",
    "    return int(est_genome_size)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "1c182bef7b0c98bc990ccdce77f3f845",
     "grade": true,
     "grade_id": "genomesize-test",
     "locked": true,
     "points": 1,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "N = 1000\n",
    "L = 36\n",
    "k = 10\n",
    "M = 30\n",
    "assert(GetGenomeSize(N, k, L, M)==900)\n",
    "M = 25\n",
    "assert(GetGenomeSize(N, k, L, M)==1080)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "24443b01ccc739a38ba18c1f407fd0c2",
     "grade": false,
     "grade_id": "text-part-realdata",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "source": [
    "## 1.3. Kmer histograms for real data\n",
    "\n",
    "In real data, our kmer histograms get slightly more complicated since we have sequencing errors. \n",
    "\n",
    "**Question 8 (1 pt)**: Modify the code below to introduce random errors into your reads at a rate of 0.1%.\n",
    "\n",
    "If your answer is correct, you should see a huge spike of unique kmers (at count=1) due to sequencing errors. In practice, to account for errors in our genome size estimation, instead of using $N$, $L$, and $k$ to compute the total number of kmers sequenced, we will instead add up all the kmers seen after chopping off the kmers with low counts that are likely due to errors. We will explore this more using real data in the lab assignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "f971d4e96c2e23d63aedb47c880440e4",
     "grade": false,
     "grade_id": "read-errors",
     "locked": false,
     "schema_version": 3,
     "solution": true
    }
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAb0AAADQCAYAAACTM790AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAAYVElEQVR4nO3dfbQdVXnH8e+PgMirIC/pNS9e1IAGXy72NoIohhchQiSggGEJK7hoQxVbUmDVhKKoGMFVULRWagQkXSoxBYEQQYWU4BsmJCEYQoikkMJt0iRaWxBLauDpH7OvHMK5907uPefMuTO/z1qzZmafeXk2JDzsmdl7KyIwMzOrgp2KDsDMzKxVnPTMzKwynPTMzKwynPTMzKwynPTMzKwynPTMzKwydi46gKHYf//9o7Ozs+gwzPJbuzZbH3JIsXGYldjy5ct/HREH1PttWCe9zs5Oli1bVnQYZvlNnJitFy8uMgqzUpP073395sebZmZWGU56ZmZWGcP68abZsOPHmmaFckvPzMwqw0nPrJWuuipbzKwQfrzZj86Z369bvv7Kk1ociZXGwoXZ+uKLi43DrKLc0jMzs8pw0jMzs8pw0jMzs8rwOz2zVtptt6IjMKs0Jz2zVrrrrqIjMKs0P940M7PKcNIza6XLL88WMytE05KepFdKWirpIUmrJX0mlb9a0t2SHkvrfWvOmSVpnaS1kk5oVmxmhVm0KFvMrBDNbOltBY6JiLcBXcAkSYcDM4FFETEOWJT2kTQemAocCkwCviZpRBPjMzOzimla0ovM79LuLmkJYAowN5XPBU5J21OAeRGxNSKeANYBE5oVn5mZVU9T3+lJGiFpJbAZuDsilgAjI2IjQFofmA4fBTxVc3pPKjMzM2uIpnZZiIjngS5J+wC3SnpzP4er3iVedpA0HZgOMHbs2EaEadY6++1XdARmldaSfnoR8d+SFpO9q9skqSMiNkrqIGsFQtayG1Nz2mhgQ51rzQHmAHR3d78sKZq1tVtuKToCs0pr5tebB6QWHpJ2A44DHgUWANPSYdOA29P2AmCqpF0lHQSMA5Y2Kz4zM6ueZrb0OoC56QvMnYD5EbFQ0v3AfEnnAk8CpwNExGpJ84FHgG3A+enxqFl5zJqVra+4otg4zCqqaUkvIn4JHFan/DfAsX2cMxuY3ayYzAp3//1FR2BWaR6RxczMKsNJz8zMKsNJz8zMKsNTC5m10ujRRUdgVmlOemat9K1vFR2BWaX58aaZmVWGk55ZK82YkS1mVgg/3jRrpZUri47ArNLc0jMzs8pw0jMzs8pw0jMzs8rwOz2zVjr44KIjMKs0Jz2zVpozp+gIzCrNjzfNzKwynPTMWmn69Gwxs0LsUNKTtK+kt+Y8doykeyWtkbRa0gWp/NOS/kPSyrScWHPOLEnrJK2VdMKOVcVsGPjVr7LFzAox4Ds9SYuBk9OxK4Etku6LiAsHOHUbcFFErJC0F7Bc0t3pty9FxFXb3Wc8MBU4FHgNcI+kgz17upmZNUqelt6rIuJp4APANyPiT4HjBjopIjZGxIq0/QywBhjVzylTgHkRsTUingDWARNyxGdmZpZLnqS3s6QO4Axg4WBuIqkTOAxYkoo+LumXkm6QtG8qGwU8VXNaD/0nSTMzsx2SJ+l9BvghsC4iHpD0OuCxvDeQtCdwCzAjtRivBV4PdAEbgat7D61zetS53nRJyyQt27JlS94wzNpDV1e2mFkh+n2nJ2kEMCYi/vjxSkQ8Dnwwz8Ul7UKW8L4dEd9L52+q+f0bvNh67AHG1Jw+Gtiw/TUjYg4wB6C7u/tlSdGsrV1zTdERmFVavy299BHJyYO5sCQB1wNrIuKLNeUdNYedCjycthcAUyXtKukgYBywdDD3NjMzqyfPiCw/l/RV4LvAs72FvR+p9ONI4GxglaSVqewS4ExJXWSPLtcD56XrrZY0H3iE7MvP8/3lppXOWWdla8+gblaIPEnvnWn92ZqyAI7p76SI+Cn139Pd2c85s4HZOWIyG556eoqOwKzSBkx6EXF0KwIxMzNrtgG/3pQ0UtL1ku5K++Mlndv80MzMzBorT5eFG8m6LLwm7f8KmNGkeMzMzJomT9LbPyLmAy8ARMQ2wB+YmA3GEUdki5kVIs+HLM9K2o/UUVzS4cD/NDUqs7K64oqiIzCrtDxJ70KyPnSvl/Qz4ADgtKZGZWZm1gR5vt5cIek9wCFkXRDWRsQfmh6ZWRl9MA1mdMstxcZhVlF5phYaAZwIdKbjj5dE7SgrZpbTb35TdARmlZbn8eYdwHPAKtLHLGZmZsNRnqQ3unbAaTMzs+EqT5eFuyQd3/RIzMzMmixPS+8XwK2SdgL+QPYxS0TE3k2NzKyMjj226AjMKi1P0rsaOAJYFRGev85sKD75yaIjMKu0PI83HwMedsIzM7PhLk9LbyOwOA04vbW30F0WzAbhfe/L1nfdVWwcZhWVp6X3BLAIeAWwV1r2HOgkSWMk3StpjaTVki5I5a+WdLekx9J635pzZklaJ2mtpBMGVyWzNva//5stZlaIPC29uRGxvrZA0p/lOG8bcFEa0WUvYLmku4FzgEURcaWkmcBM4BOSxgNTgUPJZnS4R9LBnj3dzMwaJU9L7xZJo3p3JB0F3DDQSRGxMSJWpO1ngDXAKGAKMDcdNhc4JW1PAeZFxNaIeAJYB0zIWQ8zM7MB5Ul6fwncJulPJJ0IfIVsWLLcJHUChwFLgJERsRGyxAgcmA4bBTxVc1pPKjMzM2uIPANOPyDpr4EfkQ1H9t6I2JL3BpL2BG4BZkTE05L6PLTe7etcbzowHWDs2LF5wzBrD5MnFx2BWaX1mfQk3cFLk87uZPPoXZ8GnD55oItL2oUs4X07Ir6XijdJ6oiIjZI6gM2pvAcYU3P6aGDD9teMiDnAHIDu7m53o7Dh5eKLi47ArNL6a+ldNZQLK2vSXQ+s2a57wwJgGnBlWt9eU/4dSV8k+5BlHLB0KDGYmZnV6jPpRcR9Q7z2kcDZwCpJK1PZJWTJbr6kc4EngdPT/VZLmg88Qvbl5/n+ctNKZ+LEbL14cZFRmFVWni4LgxIRP6X+ezqAugMQRsRsYHazYjIzs2rL8/WmmZlZKfSZ9CQtSusvtC4cMzOz5unv8WaHpPcAJ0uax3aPKns7npuZmQ0X/SW9T5ENETYa2H5w6QCOaVZQZqV1xhlFR2BWaf19vXkzcLOkT0bE5S2Myay8PvaxoiMwq7Q8I7JcLulk4KhUtDgiFjY3LLOS+v3vs/Xuuxcbh1lFDZj0JF1BNvDzt1PRBZKOjIhZTY3MrIxOTMPWup+eWSHy9NM7CeiKiBcAJM0FHgSc9MzMbFjJ209vn5rtVzUhDjMzs6bL09K7AnhQ0r1k3RaOwq08MzMbhvJ8yHKTpMXAn5ElvU9ExH82OzAzM7NGyzX2ZprsdUGTYzErv3POKToCs0pr2oDTZlaHk55ZoTzgtFkr/frX2WJmheg36UnaSdLDrQrGrPROOy1bzKwQ/Sa91DfvIUljd/TCkm6QtLk2aUr6tKT/kLQyLSfW/DZL0jpJayWdsKP3MzMzG0ied3odwGpJS4Fnewsj4uQBzrsR+Crwz9uVfykirqotkDQemAocCrwGuEfSwZ453czMGilP0vvMYC4cET+W1Jnz8CnAvIjYCjwhaR3Z0Gf3D+beZmZm9Qz4IUtE3AesB3ZJ2w8AQ5lL7+OSfpkef+6bykYBT9Uc05PKzMzMGmbApCfpL4Cbga+nolHAbYO837XA64EuYCNwde9t6hwbfcQzXdIyScu2bNkyyDDMCvLRj2aLmRUiz+PN88keNS4BiIjHJB04mJtFxKbebUnfAHqnKOoBxtQcOhrY0Mc15gBzALq7u+smRrO29aEPFR2BWaXl6ae3NSL+r3dH0s700QobiKSOmt1Tgd4vOxcAUyXtKukgYBywdDD3MGtrTz2VLWZWiDwtvfskXQLsJum9wMeAOwY6SdJNwERgf0k9wGXAREldZElzPXAeQESsljQfeATYBpzvLzetlM4+O1t7Pj2zQuRJejOBc4FVZEnqTuC6gU6KiDPrFF/fz/Gzgdk54jEzMxuUPLMsvJAmjl1C1kJbGxF+l2ZmZsPOgElP0knAPwH/RvaV5UGSzouIu5odnJmZWSPlebx5NXB0RKwDkPR64PuAk56ZmQ0reZLe5t6ElzwObG5SPGbldtFFRUdgVml9Jj1JH0ibqyXdCcwne6d3OtmoLGa2o97//qIjMKu0/lp6tX87NwHvSdtbgH1ffriZDWjt2mx9yCHFxmFWUX0mvYj4SCsDMauE887L1u6nZ1aIPF9vHgT8FdBZe3yOqYXMzMzaSp4PWW4j61R+B/BCU6MxMzNrojxJ77mI+ErTIzEzM2uyPEnvy5IuA34EbO0tjIihzKlnZmbWcnmS3luAs4FjePHxZqR9M9sRl15adARmlZYn6Z0KvK52eiEzG6Tjjis6ArNKyzOf3kPAPk2Ow6waVq7MFjMrRJ6W3kjgUUkP8NJ3eu6yYLajZszI1u6nZ1aIPEnvssFcWNINwGSysTvfnMpeDXyXrM/feuCMiPht+m0W2bx9zwN/HRE/HMx9zczM+jLg482IuK/ekuPaNwKTtiubCSyKiHHAorSPpPHAVODQdM7XJI3YgXqYmZkNaMCkJ+kZSU+n5TlJz0t6eqDzIuLHwH9tVzwFmJu25wKn1JTPi4itEfEEsA6YkLcSZmZmeeSZOX2v2n1JpzD4hDQyIjam626UdGAqHwX8oua4nlRmZmbWMHne6b1ERNwmaWaD41C9W9U9UJoOTAcYO3Zsg8Mwa7LPf77oCMwqLc+A0x+o2d0J6KaPhJTDJkkdqZXXwYuT0fYAY2qOGw1sqHeBiJgDzAHo7u4ebBxmxXjnO4uOwKzS8vTTe3/NcgLwDNk7uMFYAExL29OA22vKp0raNc3qMA5YOsh7mLWvn/88W8ysEHne6Q1qXj1JNwETgf0l9ZB1fbgSmC/pXOBJslnYiYjVkuYDjwDbgPMj4vnB3NesrV1ySbZ2Pz2zQvSZ9CR9qp/zIiIu7+/CEXFmHz8d28fxs4HZ/V3TzMxsKPpr6T1bp2wPsg7k+wH9Jj0zM7N202fSi4ire7cl7QVcAHwEmAdc3dd5ZmZm7arfd3pp2LALgQ+TdSZ/e++wYWZmZsNNf+/0/h74AFn3gLdExO9aFpVZWV1zTdERmFVafy29i8hmVbgU+Dvpj/3HRfYhy95Njs2srXTO/P7LytZfedKOXaSrqzHBmNmg9PdOL08fPjPbEffck609maxZIXZ4GDIzG4LPfS5bO+mZFcJJz6xg9R6bwiAenZrZgJz0zIbACctseHHSM9tOX4nMzIY/f6xiZmaV4ZaeWSt9/etFR2BWaU56Zi3U+c11aWtdv8eZWXM46Zk1QV/vBY9dtwSARW94x6Cv4Y9kzAbPSc+shf5i6a1AvqRnZo1XSNKTtJ5sBvbngW0R0Z0Gt/4u0AmsB87w4NZmZtZIRX69eXREdEVEd9qfCSyKiHHAorRvZmbWMO30eHMKMDFtzwUWA58oKhizMvB7QbOXKqqlF8CPJC2XND2VjYyIjQBpfWBBsZmZWUkV1dI7MiI2SDoQuFvSo3lPTElyOsDYsWObFZ9ZU/zN5Iuadm2PJGM2sEKSXkRsSOvNkm4FJgCbJHVExEZJHcDmPs6dQzaxLd3d3dGqmK18ikgSG/c+oOX3NLMXtTzpSdoD2CkinknbxwOfBRYA04Ar0/r2Vsdm1myT1/wYgIVvOmrQ13CLzmzwimjpjQRuTTOx7wx8JyJ+IOkBYL6kc4EngdMLiM2sqc568E5gaEnPzAav5UkvIh4H3lan/DfAsa2Ox8zMqsOzLJiZWWW0Uz89s6bwO7D83K/Pys5Jz6yC/D8CVlVOemYt9NFTZhUdglmlOemZtdBvd39V0SEMSr2WoR952nDkD1nMWui0Vfdw2qp7ig7DrLLc0rNSafd3Vb0J7+a3HFdwJEPnj15sOHJLz8zMKsNJz8zMKsNJz8zMKsNJz8zMKsMfspi10Dmnf7roEJrOH7hYO3PSM2uh53Z5ZdEhFMZ9/awd+PGmWQudteL7nLWivbtVmJWZW3rWco3oSzdcWwiTH/0JAN96+/CM32y4a7ukJ2kS8GVgBHBdRFxZcEg2SO3eUdyK5/d/1mptlfQkjQD+EXgv0AM8IGlBRDxSbGTDX9n+4+KEWm5l+/Nq7aOtkh4wAViXZldH0jxgCtBWSa8RL+Sb+VLfCcGqxh/JWF7tlvRGAU/V7PcA7ygolh3SiETTTsnK/6dtZmWkiCg6hj+SdDpwQkT8edo/G5gQEX9Vc8x0YHraPQRY24Bb7w/8ugHXaWdVqCNUo56uYzlUoY5QTD1fGxEH1Puh3Vp6PcCYmv3RwIbaAyJiDjCnkTeVtCwiuht5zXZThTpCNerpOpZDFeoI7VfPduun9wAwTtJBkl4BTAUWFByTmZmVRFu19CJim6SPAz8k67JwQ0SsLjgsMzMribZKegARcSdwZ4tv29DHpW2qCnWEatTTdSyHKtQR2qyebfUhi5mZWTO12zs9MzOzpql80pM0SdJaSeskzSw6nkaQdIOkzZIeril7taS7JT2W1vsWGeNQSRoj6V5JayStlnRBKi9NPSW9UtJSSQ+lOn4mlZemjr0kjZD0oKSFab+MdVwvaZWklZKWpbJS1VPSPpJulvRo+rt5RLvVsdJJr2bYs/cB44EzJY0vNqqGuBGYtF3ZTGBRRIwDFqX94WwbcFFEvAk4HDg//bsrUz23AsdExNuALmCSpMMpVx17XQCsqdkvYx0Bjo6IrppP+MtWzy8DP4iINwJvI/t32l51jIjKLsARwA9r9mcBs4qOq0F16wQertlfC3Sk7Q5gbdExNri+t5ON2VrKegK7AyvIRigqVR3J+uMuAo4BFqayUtUx1WM9sP92ZaWpJ7A38ATpW5F2rWOlW3rUH/ZsVEGxNNvIiNgIkNYHFhxPw0jqBA4DllCyeqbHfiuBzcDdEVG6OgLXAH8LvFBTVrY6AgTwI0nL08hSUK56vg7YAnwzPaq+TtIetFkdq570VKfMn7MOI5L2BG4BZkTE00XH02gR8XxEdJG1hiZIenPBITWUpMnA5ohYXnQsLXBkRLyd7HXK+ZKOKjqgBtsZeDtwbUQcBjxL0Y8y66h60htw2LMS2SSpAyCtNxccz5BJ2oUs4X07Ir6XiktXT4CI+G9gMdm72jLV8UjgZEnrgXnAMZK+RbnqCEBEbEjrzcCtZLPKlKmePUBPehoBcDNZEmyrOlY96VVp2LMFwLS0PY3sHdiwJUnA9cCaiPhizU+lqaekAyTtk7Z3A44DHqVEdYyIWRExOiI6yf7+/WtEnEWJ6gggaQ9Je/VuA8cDD1OiekbEfwJPSTokFR1LNi1cW9Wx8p3TJZ1I9k6hd9iz2cVGNHSSbgImko1uvgm4DLgNmA+MBZ4ETo+I/yooxCGT9C7gJ8AqXnwXdAnZe71S1FPSW4G5ZH82dwLmR8RnJe1HSepYS9JE4OKImFy2Okp6HVnrDrLHgN+JiNklrGcXcB3wCuBx4COkP7u0SR0rn/TMzKw6qv5408zMKsRJz8zMKsNJz8zMKsNJz8zMKsNJz8zMKsNJz6xJJP2uZvvENMr82CJjGipJXambj9mw5KRn1mSSjgX+AZgUEU82+NojGnm9HLoAJz0btpz0zJpI0ruBbwAnRcS/pbIbJV2b5gN8XNJ70hyIayTdWHPu8ZLul7RC0r+kcUZ752X7lKSfAqdvd7+Rkm5Nc/A9JOmdqfxCSQ+nZUYq69RL51y8WNKn0/ZiSV9QNp/fryS9O41a9FngQ2lOuA8175+cWXPsXHQAZiW2K9mQSxMj4tHtftuXbCqdk4E7yMag/HPggTSqRQ9wKXBcRDwr6RPAhWRJB+C5iHhXnXt+BbgvIk5NrcA9Jf0p2cgY7yAbZH2JpPuA3w4Q/84RMSE9zrwsIo6T9CmgOyI+vgP/HMzahlt6Zs3zB+DnwLl1frsjsuGQVgGbImJVRLwArCabC/FwsomNf5amFpoGvLbm/O/2cc9jgGvhjzM0/A/wLuDWiHg2In4HfA94d474ewfxXp5iMhv23NIza54XgDOAeyRdEhGfr/lta80xW7c7Z2fgebL5887s49rP7kAc9abQgmz2+dr/8X3ldr/3xvU8/m+FlYRbemZNFBG/ByYDH5ZUr8XXl18AR0p6A4Ck3SUdnOO8RcBH0zkjJO0N/Bg4JV1jD+BUssG6NwEHStpP0q4pzoE8A+y1A/UwaytOemZNlkaUnwRcKmlKznO2AOcAN0n6JVkSfGOOUy8Ajpa0iuyx5KERsQK4EVhKNgvFdRHxYET8gewd4RJgIdm0RQO5FxjvD1lsuPIsC2ZmVhlu6ZmZWWU46ZmZWWU46ZmZWWU46ZmZWWU46ZmZWWU46ZmZWWU46ZmZWWU46ZmZWWX8P8W7FKUK6nZ7AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 504x216 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import numpy as np\n",
    "nucs = [\"A\",\"C\",\"G\",\"T\"]\n",
    "genome_size = 1000\n",
    "GENOME = ''.join([nucs[np.random.randint(0, 4)] for  i in range(genome_size)])\n",
    "\n",
    "# Set real-world kmer size and read length\n",
    "k = 10\n",
    "L = 36\n",
    "N = 1000\n",
    "error_rate = 0.001\n",
    "\n",
    "def GetReadsWithErrors(GENOME, N, L, error_rate):\n",
    "    \"\"\" Return a list of N reads of length L from a GENOME string\n",
    "    \n",
    "    Generate random errors with probability error_rate \n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    GENOME : str\n",
    "       String consisting of the genome sequence\n",
    "    N : int\n",
    "       number of reads\n",
    "    L : int\n",
    "       length (in bp) of the reads\n",
    "    error_rate : float\n",
    "       The probability that each base is incorrect.\n",
    "       \n",
    "    Returns\n",
    "    -------\n",
    "    reads : list of str\n",
    "       List of reads (consisting of strings of nucleotides)\n",
    "    \"\"\"\n",
    "    reads = []\n",
    "    for i in range(N):\n",
    "        start = np.random.randint(0, len(GENOME)-L+1)\n",
    "        read = GENOME[start:start+L]\n",
    "        # Generate errors in the read at a per-base error rate of e.g. 0.1%\n",
    "        # Add the new read (with errors) to the list of reads\n",
    "        # The read should be a Python string\n",
    "        # your code here\n",
    "        for base in range(L):\n",
    "            if (np. random. random() < error_rate):\n",
    "                new_base = np. random.choice([x for x in nucs if x != read[base]])\n",
    "                read = read [:base] + new_base + read [base+1:]\n",
    "        #raise NotImplementedError\n",
    "        reads.append(read)\n",
    "        \n",
    "#         read_with_errors = \"\"\n",
    "#         for base in read:\n",
    "#             if np.random.random() < error_rate:\n",
    "#                 # Replace the base with a random nucleotide\n",
    "#                 read_with_errors += nucs[np.random.randint(0, 4)]\n",
    "#             else:\n",
    "#                 read_with_errors += base\n",
    "#         reads.append(read_with_errors)\n",
    "    return reads\n",
    "\n",
    "reads = GetReadsWithErrors(GENOME, N, L, error_rate)\n",
    "kmer_list = GetKmersFromReads(reads, k)\n",
    "kmer_counts = GetKmerCounts(kmer_list)\n",
    "PlotKmerHist(kmer_counts, meanval=GetMeanKmerCov(GENOME, N, k, L))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "587b61bf5946ec2b1a93ccf723905abd",
     "grade": true,
     "grade_id": "check-read-errors",
     "locked": true,
     "points": 2,
     "schema_version": 3,
     "solution": false
    }
   },
   "outputs": [],
   "source": [
    "\"\"\"Check that there are many kmers with count 1\"\"\"\n",
    "GENOME = ''.join([nucs[np.random.randint(0, 4)] for  i in range(genome_size)])\n",
    "k = 10\n",
    "L = 36\n",
    "N = 1000\n",
    "error_rate = 0.001\n",
    "reads = GetReadsWithErrors(GENOME, N, L, error_rate)\n",
    "kmer_list = GetKmersFromReads(reads, k)\n",
    "kmer_counts = GetKmerCounts(kmer_list)\n",
    "assert(len(reads) == N)\n",
    "assert(len([item for item in kmer_counts if kmer_counts[item]==1])> 50) # There is an infinitesimal chance this will fail"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "58fb5e1a390b9f6b82a0b7672bd5d3f1",
     "grade": false,
     "grade_id": "text-quizreview-intro",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "# Part 2: Review questions\n",
    "\n",
    "**Question 1 (1 pt)**: You sequence a human genome (3 billion bp) with 2 billion single end reads of 101bp each. What do you expect the average *sequencing coverage* to be? Set the variable `meancov` below to your answer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "537e0a18144fb3dcddf31b82c678cdbc",
     "grade": false,
     "grade_id": "quiz1-answer",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "meancov = 0\n",
    "# Set meancov to the average sequencing coverage\n",
    "# your code here\n",
    "g = 3e9\n",
    "n = 2e9\n",
    "l = 101\n",
    "meancov = (l*n)/g\n",
    "# raise NotImplementedError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "780810df332a883b0e0856e8f618ec38",
     "grade": true,
     "grade_id": "quiz1-test",
     "locked": true,
     "points": 1,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Check that meancov is set to the correct answer'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"Check that meancov is set to the correct answer\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "567f1d9512930a452b4a5bec0db68ae7",
     "grade": false,
     "grade_id": "cell-ef42b62b7412587c",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "**Question 2 (1 pt):** You are analyzing sequencing data of a human (which is diploid) at a position where the individual is *heterozygous* for the reference base (\"A\") and an alternate (\"T\") base. If the position was covered by 5 reads, in the absence of sequencing errors, what is the probability of seeing 0 reads with the alternate \"T\" base in that position? Set the variable `prob0` below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "ee94cc028a17ae2bcd623f41928404a0",
     "grade": false,
     "grade_id": "quiz2-answer",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.03125\n"
     ]
    }
   ],
   "source": [
    "import scipy.special\n",
    "# Recall scipy.special.comb(N, k) gives \"N choose k\" which\n",
    "# is helpful for computing binomial probabilities\n",
    "# Or you could also do the math manually and just set prob0\n",
    "# to your answer\n",
    "\n",
    "prob0 = 0\n",
    "# Set prob0 below to answer question 2\n",
    "# your code here\n",
    "prob0 = scipy.special.comb(5, 0)*(.5**0)*((1-.5)**(5-0))\n",
    "print(prob0)\n",
    "# raise NotImplementedError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "44f84ee6f697dc06030c77bc16a83c30",
     "grade": true,
     "grade_id": "quiz2-test",
     "locked": true,
     "points": 1,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "\"\"\"Check prob0 set correctly\"\"\"\n",
    "assert(prob0 >= 0 and prob0 <= 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "f2f4bc112f7649df614904c9a579af3f",
     "grade": false,
     "grade_id": "quiz3-text",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "**Question 3 (1 pt)**: You sequence a single fragment of DNA with 2x101bp paired-end Illumina reads. After aligning your reads to the genome, the first (forward) read maps to chr6:10000 and the second read maps to chr6:10400 (Note: those are the start, or left-most, coordinates of each read). What is the estimated length of the fragment based on where the reads map? Set `fraglen` to your answer below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "f82cfbd1a006e59280c699e1bf5336c2",
     "grade": false,
     "grade_id": "quiz3-answer",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "501"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fraglen = 0\n",
    "# Set fraglen to your answer below\n",
    "# your code here\n",
    "fraglen = (10400+101)-10000\n",
    "fraglen\n",
    "# raise NotImplementedError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "70aab0ceca0116cb1fc295cd20f0cac6",
     "grade": true,
     "grade_id": "quiz3-test",
     "locked": true,
     "points": 1,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "\"\"\"Check the value of fraglen\"\"\"\n",
    "assert(fraglen > 500 and fraglen < 2000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "05f34ab29e551417983d446bdf01729e",
     "grade": false,
     "grade_id": "quiz4-text",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "**Question 4 (1 pt):** You sequence a genome of length 1 billion using 3 billion single end reads of length 36bp. Using kmers of length 31, how many times do you expect to see each kmer in your data? (i.e. what is the mean *kmer coverage*)? Set your answer in the variable `kmercov` below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "2b29233f6ba1bf089dee30742212806b",
     "grade": false,
     "grade_id": "quiz4-answer",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18.000000540000016"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmercov = 0\n",
    "# Set kmercov to your answer below\n",
    "# your code here\n",
    "N= 3e9\n",
    "L = 36\n",
    "k = 31\n",
    "G= 1e9\n",
    "num_kmers_seq = N * (L - k + 1)\n",
    "num_kmers_uniq = G - k + 1\n",
    "kmercov = num_kmers_seq*1.0/num_kmers_uniq\n",
    "kmercov\n",
    "# raise NotImplementedError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "7800f5a93dd4e605c448693f6208c0e3",
     "grade": true,
     "grade_id": "quiz4-test",
     "locked": true,
     "points": 1,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "\"\"\"Check the value of kmercov\"\"\"\n",
    "assert(kmercov > 0 and kmercov < 100000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "90025cdd98ba686987445e01b9871442",
     "grade": false,
     "grade_id": "quiz5-text",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "**Question 5 (1 pt)**: Write a UNIX command to get the filesize in human readable format (e.g. number of megabytes, gigabytes, etc.) of the file `test.fq` (included in this directory). Set the variable `cmd` to your command below. Hint: you can test UNIX commands either in the terminal, or by putting `!` before your command in a Jupyter notebook cell."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "dea8af1ab1aebe4413e13597ac257f67",
     "grade": false,
     "grade_id": "quiz5-answer",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "cmd = 'echo hello world'\n",
    "# Set cmd to get the filesize in human readable format for test.fq\n",
    "# your code here\n",
    "cmd = 'ls -lh test.fq'\n",
    "# raise NotImplementedError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "deec0f77de897c5d1a13d30c15bb9b66",
     "grade": true,
     "grade_id": "quiz5-test",
     "locked": true,
     "points": 1,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "\"\"\"Check that the command gives the file size\"\"\"\n",
    "from subprocess import Popen, PIPE\n",
    "p = Popen(cmd, shell=True, stdout=PIPE, stderr=PIPE)\n",
    "out, err = p.communicate()\n",
    "assert(p.returncode == 0)\n",
    "assert(\"1.\" in str(out))\n",
    "assert(\"K\" in str(out))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "62ee5a91fbf65eee3c7615e003a563a0",
     "grade": false,
     "grade_id": "quiz6-text",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "**Question 6 (1 pt)**: Write a one line UNIX command that outputs the part of `test.fq` corresponding to reads 3-5. Remember each read encompasses 4 lines of data. Your solution should involve \"piping\" two commands together. e.g. `command1 | command2...`. Set the variable `cmd` to your command below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "bfc686e4e3203d546dc300106cd6dba9",
     "grade": false,
     "grade_id": "quiz6-answer",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "cmd = 'echo hello world'\n",
    "# your code here\n",
    "cmd =  'head -n 20 test.fq | tail -n 12'\n",
    "# raise NotImplementedError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "ceb5e9e5713a635eda12a18312014f79",
     "grade": true,
     "grade_id": "quiz6-test",
     "locked": true,
     "points": 1,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "\"\"\"Check that the command gives the correct lines\"\"\"\n",
    "from subprocess import Popen, PIPE\n",
    "p = Popen(cmd, shell=True, stdout=PIPE, stderr=PIPE)\n",
    "out, err = p.communicate()\n",
    "assert(p.returncode == 0)\n",
    "assert(len(str(out).strip().split(\"\\\\n\")[:-1])==12) # Note there is an extra \"'\" at the end. better way to write this test?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "8e6bbdd77c3e24c70cb4f7918f16ed59",
     "grade": false,
     "grade_id": "quiz7-text",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "**Question 7 (1 pt)**: Write a UNIX command to print out the first five occurrences of the string \">chr\" in the file `~/public/genomes/hg19.fa`. Set the variable `cmd` to your command below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "bce274f44240380b66a44d063dc6753b",
     "grade": false,
     "grade_id": "quiz7-answer",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "cmd = 'echo hello world'\n",
    "# Set cmd to print out the first five occurrences of the string \">chr\" \n",
    "# in the file `~/public/genomes/hg19.fa`\n",
    "# your code here\n",
    "cmd = 'grep -m 5 \">chr\" ~/public/genomes/hg19.fa'\n",
    "# raise NotImplementedError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "8dea1c1f1846da9a1367d6dcdc7161a6",
     "grade": true,
     "grade_id": "quiz7-test",
     "locked": true,
     "points": 1,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "\"\"\"Check that the command gives the correct output\"\"\"\n",
    "from subprocess import Popen, PIPE\n",
    "p = Popen(cmd, shell=True, stdout=PIPE, stderr=PIPE)\n",
    "out, err = p.communicate()\n",
    "assert(p.returncode == 0)\n",
    "assert(len(str(out).split(\"\\\\n\")[:-1])==5)\n",
    "assert(\">chr1\" in str(out))\n",
    "assert(\">chr2\" in str(out))\n",
    "assert(\">chr3\" in str(out))\n",
    "assert(\">chr4\" in str(out))\n",
    "assert(\">chr5\" in str(out))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
